File: java/src/main/java/ueb02/MVar.java
package ueb02;


public class MVar<T> {
    private T content;
    private boolean empty;

    private final Object reader = new Object();
    private final Object writer = new Object();

    public MVar(T t) {
        this.content = t;
        this.empty = false;
    }

    public MVar() {
        this.content = null;
        this.empty = true;
    }

    private boolean isEmpty() {
        return empty;
    }

    public void put(T o) throws InterruptedException {
        synchronized (writer) {
            while (!isEmpty()) {
                writer.wait();
            }

            synchronized (reader) {
                content = o;
                empty = false;
                reader.notify();
            }
        }
    }

    public T take() throws InterruptedException {
        synchronized (reader) {
            while (isEmpty()) {
                reader.wait();
            }

            synchronized (writer) {
                T result = content;
                content = null;
                empty = true;
                writer.notify();
                return result;
            }
        }
    }

    public T take(long timeOut) throws InterruptedException {
        long start = System.currentTimeMillis();

        synchronized (reader) {
            while (isEmpty()) {
                long deltaT = System.currentTimeMillis() - start;
                if (deltaT >= timeOut) {
                    return null;
                }
                reader.wait(timeOut - deltaT);
            }

            synchronized (writer) {
                T result = content;
                content = null;
                empty = true;
                writer.notify();
                return result;
            }
        }
    }

    public T read() throws InterruptedException {
        synchronized (reader) {
            while (isEmpty()) {
                reader.wait();
            }

            reader.notify();
            return content;
        }
    }

    public boolean tryPut(T toPut) {
        if (!isEmpty()) {
            return false;
        }

        synchronized (reader) {
            content = toPut;
            empty = false;
            reader.notify();
            return true;
        }
    }

    public T tryTake() {
        if (!isEmpty()) {
            synchronized (writer) {
                T result = content;

                empty = true;
                content = null;
                writer.notify();

                return result;
            }
        }

        return null;
    }

    public T swap(T toSwap) throws InterruptedException {
        synchronized (reader) {
            while (isEmpty()) {
                reader.wait();
            }

            T result = content;

            content = toSwap;
            empty = false;
            reader.notify();

            return result;
        }
    }

    public void clear() {
        synchronized (writer) {
            content = null;
            empty = true;
            writer.notify();
        }
    }

    public void overWrite(T o) {
        //todo: lol?! mind blown
    }
}

File: erlang/mvar.erl
-module(mvar).
-export([create/0, isEmpty/1, put/2, overwrite/2, take/1, take/2, read/1, tryPut/2, tryTake/1, swap/2, clear/1]).

empty() ->
	receive
		{isEmpty, Who} -> Who!{mvar_isempty, true}, empty();
		{put, V, Who} -> Who!{mvar_put, true}, full(V);
		{tryPut, V, Who} -> Who!{mvar_tryput, true}, full(V);
		{tryTake, Who} -> Who!{mvar_trytake, false}, empty();
		{overwrite, V, Who} -> Who!{mvar_overwrite, true}, full(V);
		clear -> empty()
	end.

full(C) ->
	receive
		{isEmpty, Who} -> Who!{mvar_isempty, false}, full(C);
		{take, Who} -> Who!{mvar_take, C}, empty();
		{read, Who} -> Who!{mvar_read, C}, full(C);
		{tryPut, _, Who} -> Who!{mvar_tryput, false}, full(C);
		{tryTake, Who} -> Who!{mvar_trytake, C}, empty();
		{swap, V, Who} -> Who!{mvar_swap, C}, full(V);
		{overwrite, V, Who} -> Who!{mvar_overwrite, true}, full(V);
		clear -> empty()
	end.

isEmpty(M) ->
	M!{isEmpty, self()},
	receive
		{mvar_isempty, X} -> X
	end.

put(M, V) ->
	M!{put, V, self()},
	receive
		{mvar_put, X} -> X
	end. 

overwrite(M, V) ->
	M!{overwrite, V, self()},
	receive
		{mvar_overwrite, X} -> X
	end.

take(M) ->
	M!{take, self()},
	receive
		{mvar_take, X} -> X
	end.

take(M, Ms) ->
	Self = self(),
	spawn(	fun() ->
				M!{take, self()},
				receive
					{mvar_take, X} -> Self!{mvar_take, X}
					after Ms ->
						Self!timeout
				end
			end),
	receive
		{mvar_take, X} -> X;
		timeout -> timeout
	end.


read(M) ->
	M!{read, self()},
	receive
		{mvar_read, X} -> X
	end.

tryPut(M, V) ->
	M!{tryPut, V, self()},
	receive
		{mvar_tryput, X} -> X
	end.

tryTake(M) ->
	M!{tryTake, self()},
	receive
		{mvar_trytake, X} -> X
	end.

swap(M, V) ->
	M!{swap, V, self()},
	receive
		{mvar_swap, X} -> X
	end. 

clear(M) ->
	M!clear.

create() ->
	spawn(fun() -> empty() end).